java/Mekhmet-Efe_Adzhar/src/com/zoolatech/lecture4/tasks/_2/Number2.java

```diff
@@ -19,61 +19,87 @@ public static void main(String[] args) {
         binaryTree.delete(node, 13);
         binaryTree.delete(node, 10);
         binaryTree.order(node);
+        System.out.println("ROOT NODE: " + binaryTree.getRootNodeValue());
     }
 }
 
-class BinaryTree {
+record BinaryTree(Node rootNode) {
 
-    class Node {
-        int value;
-        Node leftChild;
-        Node rightChild;
-
-        public Node(int value) {
-            this.value = value;
-        }
+    public int getRootNodeValue() {
+        return rootNode.getValue();
     }
 
     public Node insert(Node node, int value) {
         if (node == null) {
             node = new Node(value);
-        } else if (value <= node.value) {
-            node.leftChild = insert(node.leftChild, value);
+        } else if (value <= node.getValue()) {
+            node.setLeftChild(insert(node.getLeftChild(), value));
         } else {
-            node.rightChild = insert(node.rightChild, value);
+            node.setRightChild(insert(node.getRightChild(), value));
         }
         return node;
     }
 
-    void order(Node node) {
+    public void order(Node node) {
         if (node != null) {
-            order(node.leftChild);
-            System.out.println(node.value);
-            order(node.rightChild);
+            order(node.getLeftChild());
+            System.out.println(node.getValue());
+            order(node.getRightChild());
         }
     }
 
     public Node delete(Node node, int value) {
-        if (node == null) {
-            node = new Node(value);
-        }
-        if (value < node.value) {
-            node.leftChild = delete(node.leftChild, value);
-        } else if (value > node.value) {
-            node.rightChild = delete(node.rightChild, value);
+        if (value < node.getValue()) {
+            node.setLeftChild(delete(node.getLeftChild(), value));
+        } else if (value > node.getValue()) {
+            node.setRightChild(delete(node.getRightChild(), value));
         } else {
-            if (node.rightChild == null && node.leftChild == null) {
+            if (node.getRightChild() == null && node.getLeftChild() == null) {
                 node = null;
-            } else if (node.rightChild == null) {
-                node = node.leftChild;
-            } else if (node.leftChild == null) {
-                node = node.rightChild;
+            } else if (node.getRightChild() == null) {
+                node = node.getLeftChild();
+            } else if (node.getLeftChild() == null) {
+                node = node.getRightChild();
             } else {
-                Node temp = node.rightChild;
-                node.value = temp.value;
-                node.rightChild = delete(node.rightChild, temp.value);
+                Node temp = node.getRightChild();
+                node.setValue(temp.getValue());
+                node.setRightChild(delete(node.getRightChild(), temp.getValue()));
             }
         }
         return node;
     }
+}
+
+class Node {
```

Comment:
I meant class can be static. Since it represents the internal implementation of a tree, it should be defined inside of the class BinaryTree.